pkgs/purs/src/CirclesPink/Garden/Env.purs

```diff
@@ -233,6 +252,23 @@ env { request, envVars } =
     safeAddress <- CC.safePredictAddress circlesCore account { nonce: nonce }
     CC.tokenDeploy circlesCore account { safeAddress }
 
+  saveSession :: Env.SaveSession Aff
+  saveSession privKey = do
+    gundb <- liftEffect $ offline
+    _ <- liftEffect $ gundb # get "session" # put (encodeJson { privKey })
+    pure unit
+
+  restoreSession :: Env.RestoreSession Aff
+  restoreSession = do
+    gundb <- lift $ liftEffect $ offline
+    result <- gundb # get "session" # once <#> note _errRestoreSession # ExceptT
+    let
+      _ = spy "result rs" result
+    privateKey <- decodeJson result.data # lmap (const _errRestoreSession) # except
```

Comment:
@thought2 This always injects the _errRestoreSession error. It's not clear to me, how the decodeJson knows, that it should use our decodeJsonPrivateKey instance. It only has the final return type of the whole function "PrivateKey", but result.data could be any json. Do you have an idea how to fix that?

Reply:
Yes, it's correct. The compiler finds the right instance based on the types. It's defined in the types that it must return a `PrivateKey` and there can always only be one instance of a typeclass (here: DecodeJson) for a specific type (here: PrivateKey) That's why it's unambiguous and can be resolved by the compiler.

Yes, result.data can be any Json, but that's the purpose of the decoder. if the Json can be decoded to a PrivateKey it will succeed, if not it will fail.

Reply:
Ah okay! Smart compiler :)